protocols/kad: Improve options to efficiently retrieve

dignifiedquire opened the pull request:
> Still a draft as I am opening this to discuss the details and direction of this.

# Description

The end goal is to allow the following 

- when calling `get_providers` give the caller results back in real time, not collect all provider records
- when calling `get_providers` allow the total number of records to be of a fixed limit, such that not necessarily 20 peers need to be contacted (as is the case in the current impl)

# Implementation

There are currently two different commits, the first one implements the basic limit functionality, but this gets replaced with a more general solution in 2868e7bb14c9b267b2c50d811895f10cf3945b46, which streams the results and allows the caller to `.finish()` the query when they have enough.

# Things Left To Do

- [x] decide which queries should be transitioned to the new progress based api
- [x] potentially remove `Quorum` based on the new functionality, as this now could be done by the caller if using the progression based api 
- [x] write changelog

dignifiedquire committed:
feat(kad): add limit option for getting providers

dignifiedquire committed:
feat(kad): report get_providers call event based

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2701,21 +2751,26 @@ impl GetClosestPeersError {
+pub struct GetProvidersProgress {
     pub key: record::Key,
-    pub providers: HashSet<PeerId>,
+    pub provider: PeerId,
```
Why return them one by one? Why not as a batch? One always discovers them in batches, no?

dignifiedquire replied:
Yeah, I went back and for on this one. The nice thing about this, is that I can reuse the sequence number as the count of results, and the processing is a little more streamlined. But batching is definitely more efficient given they come in as batches.

thomaseizinger replied:
Do we have any idea about the difference in the performance profile?

From an API PoV, it seems nicer to return them one-by-one given that we are already returning a list of events now. For example, if I want to write a log line for each progress entry, being given a set makes this slightly more cumbersome.

dignifiedquire replied:
> Do we have any idea about the difference in the performance profile?

I don't know, the cost is `1` vs `n` events per discovery. From my understanding that cost should be quite low, but I might be mistaken.

mxinden replied:
Intuitively I am also assuming that neither has a performance benefit.

Returning them one by one does loose information. Returning them batched gives the user the information that all records within the same batch came from the same source.

dignifiedquire replied:
@mxinden updated, please check out

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2412,16 +2445,26 @@ pub enum KademliaEvent {
     },
 
+    /// An outbound query has produced a result.
+    OutboundQueryProgressed {
```
_Just a slight preference for now. Should probably put more thoughts into this._

How about the following:

- Remove `OutboundQueryProgressed`.
- Rename `OutboundQueryCompleted` to `OutboundQueryProgressed`.
- Add field `index: Index` to `OutboundQueryProgressed` (formerly `OutboundQueryCompleted`) where `enum Index { Continuing(usize), Final(usize) }` indicating the sequence number and whether it is the final update for the query.

dignifiedquire replied:
I don't have a strong preference of my solution over this, happy to change it, but lets get decide on a version before I go change it again 😅

mxinden replied:
Unless anyone of the recently tagged people (or anyone else) raises any objections, I suggest going with the above.

thomaseizinger replied:
I am unfamiliar with the internals of kademlia but do we know ahead of time, how many we will get? If yes, then always returning both numbers could be another design.

What does the number within `Final` mean? Would it work to include a boolean `completed` in the event and only have `count` otherwise?

The one thing I am worried about with your design suggestion @mxinden is that it introduces more nesting which can be cumbersome to deal with.

dignifiedquire replied:
> I am unfamiliar with the internals of kademlia but do we know ahead of time, how many we will get?

At least not in the providers case

mxinden replied:
> What does the number within `Final` mean? Would it work to include a boolean `completed` in the event and only have `count` otherwise?

Yep, that is fine as well. I want to provide two pieces of information. (1) how many of these progress updates for the same query have been emitted thus far and (2) whether this one is the final one, i.e. the query overall finished.

dignifiedquire replied:
@mxinden I have started implementing this, but running into challenges with the `QueryStats`. The current design assumes that these can be only extracted once the query is completed, but with your suggested change we would want to provide stats on every progress event. Thoughts?

mxinden replied:
> but running into challenges with the `QueryStats`. The current design assumes that these can be only extracted once the query is completed

I don't see why `QueryStats` could not be returned on each `OutboundQueryProgressed`. The `QueryStats` of a `Query` are updated on each request, thus they are always up-to-date. Am I missing something @dignifiedquire?

I am also fine returning them on the final event only, e.g. embedding them as: `Index { Continuing(usize), Final(usize, QueryStats) }` (ideally more self describing than a tuple).

dignifiedquire replied:
you were right, I was being stupid, please check out the new structure

dignifiedquire replied:
@mxinden I started expanding this to `GetRecord` and am running into issues with this construction. There is a need to differentiate the type and information between the progress events and the last event, e.g. for `GetRecord` the last event will not contain an actual record, only the optional cache. 
Thoughts?

mxinden replied:
How about instead returning the `cache_candidates` on each progression? I would expect this `BTreeMap` to be small and thus cloning to be cheap. Would you agree @dignifiedquire? Would that solve the issue?

dignifiedquire replied:
not entirely, as I would still need to make the record be optional, which is pretty confusing as a consumer

mxinden replied:
Or we return an event for each contacted peer directly with an optional record.

mxinden reviewed (commented):
I am guessing that this is already significantly speeding up your query times @dignifiedquire? Thanks for providing a patch with the discussion right away.

>  decide which queries should be transitioned to the new progress based api

Without having looked into every one, I would appreciate the consistency across the query types, thus in favor of this.

>  potentially remove Quorum based on the new functionality, as this now could be done by the caller if using the progression based api

I would be in favor of that for the sake of simplicity. I don't think anyone depends on the `Quorum` concept specifically. Not offering `Quorum` is at the expense of potentially doing extra unneeded work, e.g. when contacting another peer immediately followed by a `finish` call through a user. I doubt this is an issue though.

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2704,7 +2808,10 @@ pub type GetProvidersResult = Result<GetProvidersOk, GetProvidersError>;
+    /// The new set of providers discovered.
     pub providers: HashSet<PeerId>,
+    /// How many providers have been discovered so var.
+    pub providers_so_far: usize,
```
Why is this needed? In case a user is interested in all of the providers found, would they not need to keep track of all of them anyways?

dignifiedquire replied:
My use case was that I need it, and it is a lot easier to track in here, than to track externally. We might also need it internally if we want to do any early cancellations.

mxinden replied:
Ok. I am fine with leaving it as is.

dignifiedquire replied:
I actually realized it doesn't help me much after all, so I think we could savely remove this.

mxinden replied:
Please remove it then. Thanks.

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2472,6 +2557,25 @@ pub enum KademliaEvent {
 
+/// Information about progress events.
+#[derive(Debug, Clone)]
+pub struct Index {
```
While I proposed the name `Index`, I don't think it is intuitive.

In case anyone has more intuitive naming suggestions, that would be appreciated. //CC @thomaseizinger who usually has good ideas.

dignifiedquire replied:
I also dislike the name, but haven't had any better ideas 😓

thomaseizinger replied:
What about `ProgressStep`?

```rust
pub struct ProgressStep {
    index: usize,
    last: bool
}
```

mxinden replied:
Sounds good to me. I find `ProgressStep` a lot more intuitive than `Index`.

dignifiedquire replied:
updated

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2411,14 +2494,16 @@ pub enum KademliaEvent {
-    /// An outbound query has produced a result.
-    OutboundQueryCompleted {
+    /// An outbound query has finished.
+    OutboundQueryProgressed {
```
Is the doc comment still up to date? Should it not be "An outbound query progressed"?

dignifiedquire replied:
it is not

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2411,14 +2494,16 @@ pub enum KademliaEvent {
         /// The ID of the query that finished.
         id: QueryId,
-        /// The result of the query.
+        /// The optional final result of the query.
```
Would "The interim result of the query" not be more descriptive? I find this doc comment confusing.

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2411,14 +2494,16 @@ pub enum KademliaEvent {
         result: QueryResult,
         /// Execution statistics from the query.
         stats: QueryStats,
+        /// Indicates which event this is, if therer are multiple responses for a single request.
```
```suggestion
        /// Indicates which event this is, if therer are multiple responses for a single query.
```

I think we refer to it as _query_ everywhere else, right?

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/behaviour/test.rs:
```diff
@@ -1354,28 +1358,128 @@ fn get_providers() {
     QuickCheck::new().tests(10).quickcheck(prop as fn(_))
 }
+
+fn get_providers_limit<const N: usize>() {
```
I think our first const generic :tada:

mxinden reviewed (commented):
Thank you for making the relevant changes and thanks for propagating them across the code-base.

I left a couple of comments. Overall the direction looks great to me.

I am of the opinion that we should apply this pattern to all query types (e.g. as well to `get_closest_peers`). As far as I understand no one is objecting.

I would like to keep the `master` branch consistent at all times. Thus I would prefer the changes to the other query types to happen in this pull request as well. Would you be willing to do that work as well @dignifiedquire?

dignifiedquire committed:
change `GetRecord`api

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -680,37 +685,50 @@ where
+                })
             }
-        }
+        } else {
+            None
+        };
```
```suggestion
        }.flatten();
```

Should work as well, right?

dignifiedquire replied:
doesn't compile for me

mxinden replied:
Would require a side-effect within a `Option::map`, which in my eyes is intuitive. Sorry for the noise.

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -680,37 +685,50 @@ where
         };
         let peers = self.kbuckets.closest_keys(&target);
         let inner = QueryInner::new(info);
         let id = self.queries.add_iter_closest(target.clone(), peers, inner); // (*)
```
```suggestion
        let id = self.queries.add_iter_closest(target.clone(), peers, inner);
```

Is this reference this relevant?

dignifiedquire replied:
removed, not sure what this was from

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2538,8 +2654,8 @@ pub type GetRecordResult = Result<GetRecordOk, GetRecordError>;
 pub struct GetRecordOk {
-    /// The records found, including the peer that returned them.
-    pub records: Vec<PeerRecord>,
+    /// The record found in this iteration, including the peer that returned them.
```
```suggestion
    /// The record found in this iteration, including the peer that returned it.
```

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2133,40 +2228,39 @@ where
-                            records.push(PeerRecord {
+                            if record_to_cache.is_none() {
+                                *record_to_cache = Some(record.clone());
+                            }
```
Say we discover multiple records for the same key. With the above only the first record is replicated. `libp2p-kad` can not know which of the records is the one to be cached, i.e. it can not make a qualified judgement on which of the records is the best one.

How about removing this automated publishing mechanism and depend on the user to call `Kademlia::put_record_to` instead?

https://github.com/libp2p/rust-libp2p/blob/6da8497af5c3ce076534be68116c01b7f1748383/protocols/kad/src/behaviour.rs#L778-L798

Either way, the comment on `put_record_to` needs updating.

dignifiedquire replied:
I have removed the caching, but I am confused why the behaviour of `put_record_to` changed if I only changed the behaviour of `get_record`, does that rely on that caching to do this?

mxinden replied:
According to the Kademlia paper:

> For caching purposes, once a lookup succeeds, the requesting node stores the (key, value) pair at the closest node it observed to the key that did not return the value.

Before this pull request, when we did a `Kademlia::get_record` with a quorum of 1, we could replicate the (key, value) pair to the closest node that did not have the pair. But if we did a quorum of >1, the `Kademlia::get_record` query would potentially discover more than one value for the given key. Which value would you replicate to the closest node that did not have the value? `libp2p-kad` can not make that choice. With a quorum of >1 `Kademlia::get_record` would thus not replicate the pair to the closest node that did not have the value. Instead it relied on the user first picking the right value and then calling `Kademlia::put_record_to`.

This pull request removes the quorum mechanism. To keep things simple, I suggest also removing the automated replication to the closest node that does not have the value, but instead entirely rely on the user to call `put_record_to`.

Does that make sense @dignifiedquire? Happy to expand further.

dignifiedquire replied:
I believe https://github.com/libp2p/rust-libp2p/pull/2712/commits/d8ac894f573955c9eef914f7b832798f16e6eb3d does what you requested

mxinden reviewed (commented):
Thanks for updating this pull request refactoring `GetRecord` to the iterative approach. This looks good to me and in my eyes the consistency makes the interface easier to understand.

If I am not mistaken, from a high level, only `FindNode` is missing. In other words, I do think `GetRecord`, `GetProvider` and `FindNode` should follow the same mechanism. Do folks agree?

I don't think we need the iterative mechanism to `PutProvider` and `PutRecord` as of today. I expect these to be happening mostly async, i.e. not within a users hot path. @dignifiedquire would you agree with that intuition with iroh in mind?

dignifiedquire committed:
apply cr

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2471,6 +2547,25 @@ pub enum KademliaEvent {
+#[derive(Debug, Clone)]
+pub struct ProgressStep {
+    /// The index into the event
+    pub count: usize,
```
```suggestion
    pub count: NonZeroUsize,
```

To avoid confusion on whether this is zero-indexed or not, how about using a `NonZeroUsize` here?

dignifiedquire replied:
done

mxinden reviewed (commented):
@dignifiedquire is this ready for another review?

dignifiedquire committed:
tests and fixups for getclosestpeers

dignifiedquire committed:
remove KademliaCaching

dignifiedquire committed:
apply cr: move to nonzerousize

dignifiedquire committed:
fix example

dignifiedquire committed:
happy clippy

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2472,6 +2461,41 @@ pub enum KademliaEvent {
 
+/// Information about progress events.
+#[derive(Debug, Clone)]
+pub struct ProgressStep {
```
Needs to be exported in `kad/src/lib.rs`:

https://github.com/libp2p/rust-libp2p/blob/90df86d9e77172123562f8247f436f3e6da891e0/protocols/kad/src/lib.rs#L57-L64

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2472,6 +2461,41 @@ pub enum KademliaEvent {
+
+    pub(crate) fn nth(count: usize) -> Self {
+        Self {
+            count: NonZeroUsize::new(count).unwrap(),
```
I would prefer not unwrapping here, but instead track a `ProgressStep` internally. See cc1b2bdfc.

dignifiedquire replied:
oh sweet, thank you!

mxinden committed:
protocols/kad: Refactor step tracking

- Use `ProgressStep` to track steps internally.
- Properly report `ProgressStep` `count` on bootstrapping.

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2539,18 +2563,8 @@ pub type GetRecordResult = Result<GetRecordOk, GetRecordError>;
-    /// one of the returned records.
-    pub cache_candidates: BTreeMap<kbucket::Distance, PeerId>,
+    /// The record found in this iteration, including the peer that returned it.
+    pub record: Option<PeerRecord>,
```
I find it confusing that this is an `Option`. At least we need to document what `None` means here. Or we refactor `GetRecordOk`:

```rust
enum GetRecordOk {
  FoundRecord (PeerRecord),
  FinishedWithNoAdditionalRecord,
}
```

dignifiedquire replied:
good idea, fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2885,18 +2904,14 @@ pub enum QueryInfo {
-        /// i.e. the peers that are candidates for caching the record.
-        cache_candidates: BTreeMap<kbucket::Distance, PeerId>,
+        /// Current index of events.
+        count: usize,
```
In my eyes tracking this as a `usize` is error prone and non-descriptive. Why not use `ProgressStep` here?

Done via cc1b2bdfc in case you want to cherry-pick.

dignifiedquire replied:
fixed

mxinden commented on examples/file-sharing.rs:
```diff
@@ -326,12 +326,16 @@ mod network {
         /// Find the providers for the given file on the DHT.
         pub async fn get_providers(&mut self, file_name: String) -> HashSet<PeerId> {
-            let (sender, receiver) = oneshot::channel();
+            let (sender, mut receiver) = mpsc::channel(0);
```
Given that you only put one element into the channel below, why use a channel? Why not stick with the oneshot?

https://github.com/libp2p/rust-libp2p/pull/2712/files#diff-c0d5fae7178c7ba3a6da42df7d8c536e26fdd58762971784e3b554b7b3fc3642R454-R458

dignifiedquire replied:
reverted, this was an artificat from a different version

mxinden commented on misc/metrics/src/kad.rs:
```diff
@@ -182,7 +183,7 @@ impl super::Recorder<libp2p_kad::KademliaEvent> for Metrics {
                         Ok(ok) => self
                             .query_result_get_record_ok
-                            .observe(ok.records.len() as f64),
+                            .observe(ok.record.as_ref().map(|_| 1).unwrap_or_default() as f64),
```
Does it make sense to still track `query_result_get_record_ok` as a histogram if the only values are `0` and `1`? Wouldn't a `Counter` make more sense?

dignifiedquire replied:
yes it does, changed

mxinden commented on misc/metrics/src/kad.rs:
```diff
@@ -18,6 +18,7 @@
 // FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 // DEALINGS IN THE SOFTWARE.
 
+use libp2p_kad::GetProvidersOk;
```
All other `libp2p_kad` imports are happening in-line. I don't have an opinion on either, but I would prefer this file to be consistent.

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -446,10 +413,14 @@ where
+    pub fn replication_factor(&self) -> NonZeroUsize {
+        self.queries.config().replication_factor
+    }
+
```
Is this method necessary? We don't provide getters for any of the other config parameters.

Given that it is a static value and given that the user can set it at construction time, why is this getter needed?

dignifiedquire replied:
only needed internally, adjusted visibility

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -174,24 +171,6 @@ pub struct KademliaConfig {
-    caching: KademliaCaching,
-}
-
-/// The configuration for Kademlia "write-back" caching after successful
-/// lookups via [`Kademlia::get_record`].
-#[derive(Debug, Clone)]
-pub enum KademliaCaching {
-    /// Caching is disabled and the peers closest to records being looked up
-    /// that do not return a record are not tracked, i.e.
-    /// [`GetRecordOk::cache_candidates`] is always empty.
-    Disabled,
-    /// Up to `max_peers` peers not returning a record that are closest to the key
-    /// being looked up are tracked and returned in [`GetRecordOk::cache_candidates`].
-    /// Furthermore, if [`Kademlia::get_record`] is used with a quorum of 1, the
-    /// found record is automatically sent to (i.e. cached at) these peers. For lookups with a
-    /// quorum > 1, the write-back operation must be performed explicitly, if
-    /// desired and after choosing a record from the results, via [`Kademlia::put_record_to`].
-    Enabled { max_peers: u16 },
```
With caching removed, `PutRecordContext::Cache` needs to be removed as well.

dignifiedquire replied:
done

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -922,21 +895,44 @@ where
+                    result: QueryResult::GetProviders(Ok(GetProvidersOk {
+                        key,
+                        providers_so_far: providers.len(),
+                        providers,
```
Why return an event in case `providers` is empty? Same for `get_record` above.

dignifiedquire replied:
you are totally right, fixed this

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -1455,37 +1437,43 @@ where
                         peer,
                         num_remaining,
                     })),
+                    step: ProgressStep::single(),
```
There might have been `OutboundQueryProgressed` events emitted for this `bootstrap` call.

mxinden replied:
Would be fixed with https://github.com/libp2p/rust-libp2p/commit/cc1b2bdfc09a89a56318e38861204ed2c204cfb8.

dignifiedquire replied:
should be fixed

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2704,7 +2714,10 @@ pub type GetProvidersResult = Result<GetProvidersOk, GetProvidersError>;
     pub providers: HashSet<PeerId>,
+    /// How many providers have been discovered so var.
+    pub providers_so_far: usize,
     pub closest_peers: Vec<PeerId>,
```
Is `closest_peers` helpful on any of the `OutboundQueryProgressed` but the very last one? Maybe we can split `GetProvidersOk` via an enum similar to `GetRecordOk`:

``` rust 
enum GetProvidersOk {
  FoundProviders {
    ...
  },
  FinishedWithNoAdditionalRecord {
    closest_peers: Vec<PeerId>,
  }
}
```

If I am not mistaken we currently also emit a `OutboundQueryProgressed` with an empty `providers` in `query_finished`. With the above, this would no longer need to be the case.

dignifiedquire replied:
yes, very good idea, done

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2539,18 +2563,8 @@ pub type GetRecordResult = Result<GetRecordOk, GetRecordError>;
 pub struct GetRecordOk {
-    /// The records found, including the peer that returned them.
-    pub records: Vec<PeerRecord>,
-    /// If caching is enabled, these are the peers closest
-    /// _to the record key_ (not the local node) that were queried but
-    /// did not return the record, sorted by distance to the record key
-    /// from closest to farthest. How many of these are tracked is configured
-    /// by [`KademliaConfig::set_caching`]. If the lookup used a quorum of
-    /// 1, these peers will be sent the record as a means of caching.
-    /// If the lookup used a quorum > 1, you may wish to use these
-    /// candidates with [`Kademlia::put_record_to`] after selecting
-    /// one of the returned records.
-    pub cache_candidates: BTreeMap<kbucket::Distance, PeerId>,
```
I think there was a misunderstanding here.

> This pull request removes the quorum mechanism. To keep things simple, I suggest also removing the automated replication to the closest node that does not have the value, but instead entirely rely on the user to call `put_record_to`.

`https://github.com/libp2p/rust-libp2p/pull/2712/#discussion_r984483930`

You removed the automated caching in `query_finished`. That is good. We now rely on the user to call `put_record_to`. But given that you remove the `cache_candidates` on the `GetRecordOk` `struct`, the user no longer knows with which peers it is supposed to call `put_record_to`. We still need to track and emit `cache_candidates`.

Does the above make sense?

dignifiedquire replied:
Yes very much, my mistake entirely. I brought back `KademliaCaching` and `cache_candidates` to enable tracking this now.

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2854,14 +2866,21 @@ pub enum QueryInfo {
     },
 
     /// A query initiated by [`Kademlia::get_closest_peers`].
```
```suggestion
    /// A (repeated) query initiated by [`Kademlia::get_closest_peers`].
```

To be consistent with the comment updates below.

dignifiedquire replied:
fixed

mxinden reviewed (commented):
Went through all of the changes. In my eyes there is still work to do to make the API changes consistent.

dignifiedquire committed:
cr feedback round 1

dignifiedquire committed:
cr: improve api for GetRecordOk and GetProvidersOk

dignifiedquire committed:
fixup rebase of examples

dignifiedquire committed:
switch to counter for number of observed record

dignifiedquire committed:
bring back kademliacaching

mxinden commented on examples/file-sharing.rs:
```diff
@@ -331,7 +331,11 @@ mod network {
+            if let Ok(h) = receiver.await {
+                out.extend(h);
+            }
+            out
```
`h` is already a `HashSet`, why create a new one and then extend it?

mxinden replied:
See 1398ee6a6.

dignifiedquire replied:
@mxinden I am not seeing that commit anywhere, did you push this?

mxinden replied:
My bad. Updated my branch now:

https://github.com/mxinden/rust-libp2p/commits/feat-kad-count

Unfortunately the hashes above don't link. I added the comments before I pushed the commits.

dignifiedquire replied:
thanks, what a service, getting suggestions already commited <3

mxinden commented on examples/file-sharing.rs:
```diff
@@ -440,6 +448,19 @@ mod network {
+                        .pending_get_providers
+                        .remove(&id)
+                        .expect("Completed query to be previously pending.");
+                }
```
How about calling `Kademlia::query_mut()::finish()` when discovering the first provider record? We found one record, we are not interested in any additional records.

With this suggestion the second match on `QueryResult::GetProviders` would only need to match on `GetProvidersOk::FinishedWithNoAdditionalRecord`.

mxinden replied:
See f8d36f23c.

dignifiedquire replied:
merged, 🙏

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -446,10 +445,14 @@ where
     }
 
+    /// Returns the currently configured `replication_factor`.
+    pub(crate) fn replication_factor(&self) -> NonZeroUsize {
```
```suggestion
    fn replication_factor(&self) -> NonZeroUsize {
```

Why does this need to be `pub(crate)`? Tests succeed with this diff on my machine.

mxinden replied:
See 24f54bb90.

dignifiedquire replied:
merged, 🙏

mxinden committed:
examples/file-sharing: Revert usage of HashSet

mxinden commented on examples/file-sharing.rs:
```diff
@@ -440,6 +448,19 @@ mod network {
+                    // Drop channel to signal query is complete.
+                    let _ = self
+                        .pending_get_providers
+                        .remove(&id)
```
Given the `remove` a couple of lines above, wouldn't this panic?

mxinden replied:
See f8d36f23c.

dignifiedquire replied:
merged, 🙏

mxinden committed:
examples/file-sharing: Finish query once provider is found

mxinden committed:
protocols/kad: Remove pub(crate) from replication_factor

mxinden committed:
protocols/kad: Refactor get_record

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -680,36 +685,49 @@ where
-            records,
-            quorum,
+            step: ProgressStep::first(),
+            found_a_record: record.is_some(),
```
I find it confusing that `step` is later on altered based on `record` and `found_a_record` is altered based on `record` right away.

15b30a25d would fix that.

dignifiedquire replied:
merged, 🙏

mxinden committed:
protocols/kad: Refactor get_providers step instantiation

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -922,21 +938,47 @@ where
+            let query = self.query_mut(&id).expect("just inserted");
+            if let QueryInfo::GetProviders { step, .. } = &mut query.query.inner.info {
+                *step = step.next();
+            }
```
Same here. Altering this after creation is confusing to me. I suggest to do the same change as done for `get_record`. See 5e8f3cce8.

dignifiedquire replied:
merged, 🙏

mxinden committed:
protocols/kad: Remove pub from ProgressStep methods

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2472,6 +2523,43 @@ pub enum KademliaEvent {
+
+impl ProgressStep {
+    // TODO Why the pub?
+    pub(crate) fn first() -> Self {
```
I don't think `pub(crate)` is needed. See 0f0399bae.

dignifiedquire replied:
merged, 🙏

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2702,10 +2787,15 @@ pub type GetProvidersResult = Result<GetProvidersOk, GetProvidersError>;
+        providers: HashSet<PeerId>,
+    },
+    FinishedWithNoAdditionalRecord {
+        closest_peers: Vec<PeerId>,
```
As an aside, while we have a `put_record_to` to put a record to an individual node, we don't have a `put_provider_record_to`. 

That seems surprising to me, but I guess not needed by anyone out there thus far.

(Not to be fixed in this pull request.)

dignifiedquire replied:
👍

mxinden commented on protocols/kad/src/behaviour/test.rs:
```diff
@@ -801,7 +803,6 @@ fn get_record() {
 
 #[test]
 fn get_record_many() {
-    // TODO: Randomise
```
Why should we no longer randomise this value?

dignifiedquire replied:
readded

mxinden commented on protocols/kad/src/behaviour/test.rs:
```diff
@@ -1352,30 +1366,137 @@ fn get_providers() {
+
+#[test]
+fn get_providers_limit_n_2() {
+    get_providers_limit::<1>();
```
```suggestion
    get_providers_limit::<2>();
```

dignifiedquire replied:
fixed

mxinden commented on protocols/kad/src/query.rs:
```diff
@@ -419,6 +419,18 @@ impl<TInner> Query<TInner> {
+            }
+            QueryPeerIter::Fixed(ref iter) => Either::Right(iter.as_intermediary_result()),
+        }
+    }
```
Given that we only return the closest nodes on the alst `QueryProgressed`, this is no longer called, right? Mind removing it?

dignifiedquire replied:
removed

mxinden commented on protocols/kad/src/query/peers/closest.rs:
```diff
@@ -365,6 +365,19 @@ impl ClosestPeersIter {
+            })
+            .take(self.config.num_results.get())
+    }
+
```
Same here.

dignifiedquire replied:
removed

mxinden commented on protocols/kad/src/query/peers/closest/disjoint.rs:
```diff
@@ -331,6 +331,15 @@ impl ClosestDisjointPeersIter {
+            .map(|iter| iter.clone().into_result().map(Key::from));
+
+        ResultIter::new(self.target.clone(), result_per_path).map(Key::into_preimage)
+    }
```
And here.

dignifiedquire replied:
removed

mxinden commented on protocols/kad/src/query/peers/fixed.rs:
```diff
@@ -170,6 +170,16 @@ impl FixedPeersIter {
+                None
+            }
+        })
+    }
```
And this one.

dignifiedquire replied:
removed

mxinden reviewed (commented):
Thanks for applying my suggestions. This is likely the last large review.

Most (not all) of the new suggestions are fixed on my branch. If you like them all, you can just merge my branch into yours.

https://github.com/mxinden/rust-libp2p/tree/feat-kad-count

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -446,10 +445,14 @@ where
         }
     }
 
+    /// Returns the currently configured `replication_factor`.
```
```suggestion
```

Thinking about this some more, this is only needed in tests. In those tests, we can get the value from the `Default` implementation of `KademliaConfig`. 

I don't think we should introduce a public facing method just for testing.

dignifiedquire replied:
fixed

dignifiedquire committed:
remove unused as_intermediary_result

dignifiedquire committed:
fix test cr comments

dignifiedquire committed:
fixup: rebase

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2492,6 +2547,42 @@ pub enum KademliaEvent {
+            count: self
+                .count
+                .checked_add(1)
+                .expect("Not to exhaust usize space."),
```
```suggestion
                // TODO: Use checked_add with Rust 1.64.
                .saturating_add(1),
```

This is only in stable with `1.64`.

```
error[E0658]: use of unstable library feature 'nonzero_ops'
    --> protocols/kad/src/behaviour.rs:2579:18
     |
2579 |                 .checked_add(1)
     |                  ^^^^^^^^^^^
     |
     = note: see issue #84186 <https://github.com/rust-lang/rust/issues/84186> for more information
     ```

dignifiedquire replied:
`saturating_add` is only available on `1.64` as well: https://doc.rust-lang.org/std/num/struct.NonZeroUsize.html#method.saturating_add 

Fixed it in an ugly way..

thomaseizinger replied:
> `saturating_add` is only available on `1.64` as well: https://doc.rust-lang.org/std/num/struct.NonZeroUsize.html#method.saturating_add 
> 
> Fixed it in an ugly way.. 

I don't mind bump MSRV as long as it is documented in the changelog :)

dignifiedquire replied:
seems not really worth imho for this

dignifiedquire committed:
use get instead of checked_add

mxinden commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2492,6 +2547,37 @@ pub enum KademliaEvent {
+
+    fn next(&self) -> Self {
+        assert!(!self.last);
+        let count = NonZeroUsize::new(self.count.get() + 1).expect("Not to exhaust usize space.");
```
Nit pick: I don't think `+` in Rust overflows but instead panics, thus the `expect` is actually not checking for `exhaust`.

``` rust
#[allow(arithmetic_overflow)]

fn main() {
    assert_eq!(255_u8 + 1, 0);
}
```

```
Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 1.61s
     Running `target/debug/playground`
thread 'main' panicked at 'attempt to add with overflow', src/main.rs:4:16
note: [run with `RUST_BACKTRACE=1` environment variable to display a backtrace](https://play.rust-lang.org/#)
```

```suggestion
        let count = NonZeroUsize::new(self.count.get() + 1).expect("Adding 1 not to result in 0.");
```

dignifiedquire replied:
It's a bit more complicated, overflow checks are only inserted in `debug` builds, in `release` builds this will wrap around ( and the above program you posted will pass)

dignifiedquire committed:
Update protocols/kad/src/behaviour.rs

Co-authored-by: Max Inden <mail@max-inden.de>

dignifiedquire committed:
Merge branch 'master' into feat-kad-count

mxinden reviewed (commented):
Great. This looks good to me from a code perspective. Only missing piece would be a changelog entry, allowing folks unfamiliar with this change to upgrade their downstream code easily. Things to cover:

- Rename to `OutboundQueryProgressed`.
- Mention that multiple events per query are emitted.
- Mention that one needs to manually finish a query in case one event is enough.
- ...

dignifiedquire committed:
Merge remote-tracking branch 'upstream/master' into feat-kad-count

dignifiedquire committed:
add changelog

mxinden commented on protocols/kad/CHANGELOG.md:
```diff
@@ -16,10 +16,18 @@
 
+- Refactor APIs to be streaming.
+  - Renamed `KademliaEvent::OutboundQueryCompleted` to `KademliaEvent::OutboundQueryProgressed`
+  - Instead of a single event `OutboundQueryCompleted`, there are now multiple events emitted, allowing the user to process them as they come in (via the new `OutboundQueryProgressed`).
```
```suggestion
  - Instead of a single event `OutboundQueryCompleted`, there are now multiple events emitted, allowing the user to process them as they come in (via the new `OutboundQueryProgressed`). See `ProgressStep` to identify the final `OutboundQueryProgressed` of a single query.
```

mxinden commented on protocols/kad/CHANGELOG.md:
```diff
@@ -16,10 +16,18 @@
+- Refactor APIs to be streaming.
+  - Renamed `KademliaEvent::OutboundQueryCompleted` to `KademliaEvent::OutboundQueryProgressed`
+  - Instead of a single event `OutboundQueryCompleted`, there are now multiple events emitted, allowing the user to process them as they come in (via the new `OutboundQueryProgressed`).
+  - `get_providers` and `get_record` do not terminate automatically anymore, the caller needs to call `query.finish()`, on termination conditions.
```
```suggestion
  - To finish a query early, i.e. before the final `OutboundQueryProgressed` of the query, a caller needs to call `query.finish()`.
```

They do terminate automatically. Though not before the final `OutboundQueryProgressed`. The above suggestion would make that easier to understand for me.

mxinden commented on protocols/kad/CHANGELOG.md:
```diff
@@ -16,10 +16,18 @@
+  - Renamed `KademliaEvent::OutboundQueryCompleted` to `KademliaEvent::OutboundQueryProgressed`
+  - Instead of a single event `OutboundQueryCompleted`, there are now multiple events emitted, allowing the user to process them as they come in (via the new `OutboundQueryProgressed`).
+  - `get_providers` and `get_record` do not terminate automatically anymore, the caller needs to call `query.finish()`, on termination conditions.
+  - There is no more automatic caching of records, the user has to manually call `put_record_to` on the `QueryInfo::GetRecord.cache_candiates`.
```
```suggestion
  - There is no more automatic caching of records. The user has to manually call `put_record_to` on the `QueryInfo::GetRecord.cache_candidates` to cache a record to a close peer that did not return the record on the foregone query.
```

mxinden reviewed (commented):
Some small suggestions. Otherwise this ready to go from my end.

dignifiedquire committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Max Inden <mail@max-inden.de>

dignifiedquire committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Max Inden <mail@max-inden.de>

dignifiedquire committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Max Inden <mail@max-inden.de>

dignifiedquire committed:
Merge branch 'master' into feat-kad-count

mxinden reviewed (approved):
Looks good to me. Thanks for bearing with me @dignifiedquire!
